### DISCUSSION TITLE
JWT TOKEN

### DISCUSSION THREAD

**Discussion by marijang** [nextauthjs/next-auth#10232] [Help]
Hey, 

I have already java spring boot rest services and i want use them.
Rest services is with jwt token.

Can I or what is the best way to connect social login like google or facebook with java spring rest?
Can i use the same secret?
How to get jwt token in session data?

I m using latest version? "beta"

**Comment by mvn-h24** [comment]
> How to get jwt token in session data?

You can use [hook](https://next-auth.js.org/getting-started/client#usesession) but before you need setting up next-auth config, for example take [answer](https://github.com/nextauthjs/next-auth/discussions/10120#discussioncomment-8574826) with CredentialsProvider. Logic will be same but not provider.


> I m using latest version? "beta"

I using [this package ](https://www.npmjs.com/package/next-auth) latest version and it works


> Can i use the same secret?

Yes

> Can I or what is the best way to connect social login like google or facebook with java spring rest?

Use [providers](https://next-auth.js.org/providers/google)  in config.
For example, in GoogleProvider with `profile` callback able to make request for backend with `id_token` and signin user in backend and response tokens for frontend
```typescript
GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID ?? "",
      clientSecret: process.env.GOOGLE_CLIENT_SECRET ?? "",
      profile: (profile: any, set: TokenSet) => {
        const { picture } = profile;
        const { id_token } = set;
        if (id_token === undefined) {
          throw new Error("sign in error, not found id_token");
        }
        return signInGoogle(id_token).then(
          ({
            data: {
              user: { id, email, name },
              access_token,
              refresh_token,
            },
          }) => {
            const res = {
              id,
              email,
              image: picture,
              name,
              tokenPair: {
                access: access_token,
                refresh: refresh_token,
              },
            };
            console.log('signin', res);
            return res;
          }
        );
      },
    }),
```

**Reply by marijang** [reply]
hm....

So basicly I want this:

1. On Nextjs and Next auth user hits google social login
2. After success i need to match user on my spring boot rest services because spring boot access jwt token is different ...so basicly i need some mechanisam to exchange token for mine....
3. When i put your code i have errors: 
     - Cannot find function signInGoogle
     - Cannot find TokenSet
     - Google provider in new version is maybe just Google

**Reply by mvn-h24** [reply]
1. On frontend you need just call `signIn('google')` [method](https://next-auth.js.org/getting-started/client#signin)
2. After api call from `profile()` callback you need to configure other [callbacks](https://next-auth.js.org/configuration/callbacks) 
2.1. `signIn` , prepare custom backend data for next callbacks
2.2. `jwt` pass data which will be included into token
2.3. `session` to control data of your session
So in cookies of frontend app your token will be in httponly token and for decades this next-author makes self-request by useSession hook 
3. Yes because i took this piece from context of my app.

- `signInGoogle` it will be your api-request to spring boot
- `import NextAuth, { NextAuthOptions, TokenSet } from "next-auth";`
- `import GoogleProvider from "next-auth/providers/google";` _default imports will make babylon of contexts_

✌️ ❤️ 🌴

**Reply by marijang** [reply]
which version are u using? i m using latest and there is not Tokenset exported.
